README.md
opengalax - touchscreen daemon
==============================

&copy; 2012 Pau Oliva Fora - pof[at]eslack(.)org

Opengalax is a Linux userland input device driver for touchscreen panels manufactured by [EETI](http://home.eeti.com.tw/web20/eGalaxTouchDriver/linuxDriver.htm): eGalax, eMPIA, TouchKit, Touchmon, HanTouch, D-WAV Scientific Co, eTurboTouch, Faytech, PanJit, 3M MicroTouch EX II, ITM, etc.

The panels can be connected using 3 different interfaces: 
- USB (Hardware ID: USB\VID_0EEF&PID_0001 or USB\VID_0EEF&PID_0002)
- RS232 (Hardware ID: SERNUM\EGX5800, SERNUM\EGX5900, SERNUM\EGX6000, SERNUM\EGX5901 and SERNUM\EGX5803)
- PS/2 (Hardware ID: *PNP0F13)

At the moment only PS/2 (via serio_raw) interface is supported because this is the hardware I have, but it shouldn't be
difficult to adapt the source for USB or Serial (RS232) devices. Feel free to fork and send a pull request if you can
adapt it for other devices/interfaces.

**Why?** Because EETI only offers closed source binary drivers for those touch panels, the eGalax Touch driver is outdated
and doesn't work properly on new Xorg servers (the module ABI differs), and the newer closed source eGTouch daemon driver
doesn't work properly with PS2 devices like mine, so I wrote opengalax to have an alternative Open Source (GPL) driver.


Configuration
-------------

When first launched opengalax will create a configuration file in `/etc/opengalax.conf' with the default configuration values:

    # opengalax configuration file

    #### config data:
    serial_device=/dev/serio_raw0
    uinput_device=/dev/uinput
    rightclick_enable=0
    rightclick_duration=350
    rightclick_range=10
    # direction: 0 = normal, 1 = invert X, 2 = invert Y, 4 = swap X with Y
    direction=0
    # set psmouse=1 if you have a mouse connected into the same port
    # this usually requires i8042.nomux=1 and i8042.reset kernel parameters
    psmouse=0

    #### calibration data:
    # - values should range from 0 to 2047
    # - right/bottom must be bigger than left/top
    # left edge value:
    xmin=0
    # right edge value:
    xmax=2047
    # top edge value:
    ymin=0
    # bottom edge value:
    ymax=2047


When launched without parameters, opengalax will read the configuration from this
config file, some configuration values can also be overwritten via the command line:

    Usage: opengalax [options]
        -c                   : calibration mode
    	-f                   : run in foreground (do not daemonize)
    	-s <serial-device>   : default=/dev/serio_raw0
    	-u <uinput-device>   : default=/dev/uinput


Calibration
-----------

Altough opengalax provides a basic calibration mode (-c command line switch), for best results
it is recommended to use xinput_calibrator and leave the default values in opengalax configuration file.

Usage in Xorg
-------------

Opengalax will configure evdev xorg driver to handle right click emulation by default, so you don't need
to enable right click emulation in the configuration file. If for some reason you do not want to use evdev,
opengalax can handle the right click emulation itself by enabling it in the configuration file.

Ubuntu packages
---------------
Official Ubuntu packages are available in [poliva/opengalax ppa](https://launchpad.net/~poliva/+archive/opengalax):

     sudo add-apt-repository ppa:poliva/opengalax
     sudo apt-get update
     sudo apt-get install opengalax

debian/changelog
opengalax (0.4-ubuntu1) precise; urgency=low

  * Initial release.

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Mon, 02 Jul 2012 20:18:33 +0200


psmouse.c
/*
 *   opengalax touchscreen daemon
 *   Copyright 2012 Pau Oliva Fora <pof@eslack.org>
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License
 *   as published by the Free Software Foundation; either version 2
 *   of the License, or (at your option) any later version.
 *
 *
 * PS/2 mouse driver
 * 
 * Copyright (c) 1999-2002 Vojtech Pavlik
 * Copyright (c) 2004 Sau Dan LEE
 */


#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/time.h>
#include <stdlib.h>
#include <linux/input.h>
#include <linux/uinput.h>
#include "psmouse.h"
#include "opengalax.h"


static unsigned int psmouse_max_proto = -1U;
int psmouse_resolution = 200;
unsigned int psmouse_rate = 100;
static const char *psmouse_protocols[] = { "None", "PS/2", "PS2++", "PS2T++", "GenPS/2", "ImPS/2", "ImExPS/2", "SynPS/2"};

static struct psmouse psmouse_data;
struct psmouse * const psmouse = &psmouse_data;

/********** functions for interacting with 'uinput' **********/

static int psmouse_uinput_fd;

void uinput_open(const char *uinput_dev_name) {
	psmouse_uinput_fd = open(uinput_dev_name, O_WRONLY);
}

void uinput_set_evbit(int bit) {
	int r;

	r = ioctl(psmouse_uinput_fd, UI_SET_EVBIT, bit);
	if (r==-1) { pferrx(); }
}

void uinput_set_ledbit(int bit) {
	int r;

	r = ioctl(psmouse_uinput_fd, UI_SET_LEDBIT, bit);
	if (r==-1) { pferrx();}
}

void uinput_set_keybit(int bit) {
	int r;

	r = ioctl(psmouse_uinput_fd, UI_SET_KEYBIT, bit);
	if (r==-1) { pferrx(); }
}

void uinput_set_relbit(int bit) {
	int r;

	r = ioctl(psmouse_uinput_fd, UI_SET_RELBIT, bit);
	if (r==-1) { pferrx(); }
}


void uinput_create() {
	struct uinput_user_dev uinput = {
		.name = "psmouse ",
	};
	int r;

	snprintf(uinput.name, UINPUT_MAX_NAME_SIZE, "psmouse");

	r = write(psmouse_uinput_fd, &uinput, sizeof(uinput));
	if (r==-1) { pferrx(); }

	r = ioctl(psmouse_uinput_fd, UI_DEV_CREATE);
	if (r==-1) { pferrx(); }
}

void uinput_destroy() {
	int r;

	r = ioctl(psmouse_uinput_fd, UI_DEV_DESTROY);
	if (r==-1) { pferrx(); }
}

void uinput_close() {
	int r;

	r = close(psmouse_uinput_fd);
	if (r==-1) { pferrx(); }
}

void uinput_event(__u16 type, __u16 code, __s32 value) {
	struct input_event ev = {
		.type = type,
		.code = code,
		.value = value
	};
	int r;

	r = write(psmouse_uinput_fd, &ev, sizeof(ev));
	if (r!=sizeof(ev)) { pferrx(); }
}



/********** functions for accessing the raw keyboard device **********/

static int phys_write(unsigned char byte) {
	int r;
	r = write(fd_serial, &byte, 1);
	if (r==-1) { pferrx(); }
	if (r!=1) { err("cannot write"); exit(1); }
	return 0;
}

void psmouse_interrupt(unsigned char data);

int phys_wait_for_input(int *ptimeout) {
	unsigned char byte;
	int r;

	if (ptimeout != NULL) {
		fd_set fds;
		struct timeval tv;

		FD_ZERO(&fds);
		FD_SET(fd_serial, &fds);
		tv.tv_sec=0; tv.tv_usec=*ptimeout;
		r = select(fd_serial+1, &fds, NULL, &fds, &tv);
		*ptimeout = tv.tv_sec*1000000 + tv.tv_usec;
		if (r==-1) { pferrx(); }
		if (r==0) { return -1; } /* timeout */
		if (r!=1) { err("cannot select"); exit(1); }
	}

	r = read(fd_serial, &byte, 1);
	if (r==-1) { pferrx(); }
	if (r!=1) { err("cannot read"); exit(1); }

	psmouse_interrupt(byte);
	return 0;
}

static int psmouse_reconnect();

static void phys_rescan() {
	psmouse_disconnect();
	psmouse_connect();
}

void phys_reconnect() {
	psmouse_reconnect();
}



/*
 * psmouse_process_packet() analyzes the PS/2 mouse packet contents and
 * reports relevant events to the input module.
 */

static void psmouse_process_packet()
{
	unsigned char *packet = psmouse->packet;

/*
 * Generic PS/2 Mouse
 */

	uinput_report_key(BTN_LEFT,    packet[0]       & 1);
	uinput_report_key(BTN_MIDDLE, (packet[0] >> 2) & 1);
	uinput_report_key(BTN_RIGHT,  (packet[0] >> 1) & 1);

	uinput_report_rel(REL_X, packet[1] ? (int) packet[1] - (int) ((packet[0] << 4) & 0x100) : 0);
	uinput_report_rel(REL_Y, packet[2] ? (int) ((packet[0] << 3) & 0x100) - (int) packet[2] : 0);

	uinput_sync();
}

/*
 * psmouse_interrupt() handles incoming characters, either gathering them into
 * packets or passing them to the command routine as command output.
 */

void psmouse_interrupt(unsigned char data)
{
	if (psmouse->state == PSMOUSE_IGNORE)
		goto out;

	if (psmouse->acking) {
		switch (data) {
			case PSMOUSE_RET_ACK:
				psmouse->ack = 1;
				break;
			case PSMOUSE_RET_NAK:
				psmouse->ack = -1;
				break;
			default:
				psmouse->ack = 1;	/* Workaround for mice which don't ACK the Get ID command */
				if (psmouse->cmdcnt)
					psmouse->cmdbuf[--psmouse->cmdcnt] = data;
				break;
		}
		psmouse->acking = 0;
		goto out;
	}

	if (psmouse->cmdcnt) {
		psmouse->cmdbuf[--psmouse->cmdcnt] = data;
		goto out;
	}

	{
	struct timeval tv;
	long long jiffies;
	int r;

	r = gettimeofday(&tv, NULL);
	if (r==-1) { pferrx(); }
	jiffies = tv.tv_sec * 1000000 + tv.tv_usec;

	if (psmouse->state == PSMOUSE_ACTIVATED &&
	    psmouse->pktcnt 
	    &&  jiffies > psmouse->last + 500000) {
		warn("%s lost synchronization, throwing %d bytes away.\n",
		       psmouse->name, psmouse->pktcnt);
		psmouse->pktcnt = 0;
	}

	psmouse->last = jiffies;
	}

	psmouse->packet[psmouse->pktcnt++] = data;

	if (psmouse->packet[0] == PSMOUSE_RET_BAT) {
		if (psmouse->pktcnt == 1)
			goto out;

		if (psmouse->pktcnt == 2) {
			if (psmouse->packet[1] == PSMOUSE_RET_ID) {
				psmouse->state = PSMOUSE_IGNORE;
				phys_rescan();
				goto out;
			}
		}
	}

	if (psmouse->pktcnt == 3 + (psmouse->type >= PSMOUSE_GENPS)) {
		psmouse_process_packet();
		psmouse->pktcnt = 0;
		goto out;
	}
out:
	return;
}

/*
 * psmouse_sendbyte() sends a byte to the mouse, and waits for acknowledge.
 * It doesn't handle retransmission, though it could - because when there would
 * be need for retransmissions, the mouse has to be replaced anyway.
 */

static int psmouse_sendbyte(unsigned char byte)
{
	int timeout = 10000; /* 100 msec */
	psmouse->ack = 0;
	psmouse->acking = 1;

	if (phys_write(byte)) {
		psmouse->acking = 0;
		return -1;
	}

	while (!psmouse->ack && timeout>0)
		phys_wait_for_input(&timeout);

	return -(psmouse->ack <= 0);
}

/*
 * psmouse_command() sends a command and its parameters to the mouse,
 * then waits for the response and puts it in the param array.
 */

int psmouse_command(unsigned char *param, int command)
{
	int timeout = 500000; /* 500 msec */
	int send = (command >> 12) & 0xf;
	int receive = (command >> 8) & 0xf;
	int i;

	psmouse->cmdcnt = receive;

	if (command == PSMOUSE_CMD_RESET_BAT)
                timeout = 4000000; /* 4 sec */

	/* initialize cmdbuf with preset values from param */
	if (receive)
	   for (i = 0; i < receive; i++)
		psmouse->cmdbuf[(receive - 1) - i] = param[i];

	if (command & 0xff)
		if (psmouse_sendbyte(command & 0xff))
			return (psmouse->cmdcnt = 0) - 1;

	for (i = 0; i < send; i++)
		if (psmouse_sendbyte(param[i]))
			return (psmouse->cmdcnt = 0) - 1;

	while (psmouse->cmdcnt && timeout>0) {

		if (psmouse->cmdcnt == 1 && command == PSMOUSE_CMD_RESET_BAT &&
				timeout > 100000) /* do not run in a endless loop */
			timeout = 100000; /* 1 sec */

		if (psmouse->cmdcnt == 1 && command == PSMOUSE_CMD_GETID &&
		    psmouse->cmdbuf[1] != 0xab && psmouse->cmdbuf[1] != 0xac) {
			psmouse->cmdcnt = 0;
			break;
		}

		phys_wait_for_input(&timeout);
	}

	for (i = 0; i < receive; i++)
		param[i] = psmouse->cmdbuf[(receive - 1) - i];

	if (psmouse->cmdcnt)
		return (psmouse->cmdcnt = 0) - 1;

	return 0;
}


/*
 * psmouse_reset() resets the mouse into power-on state.
 */
int psmouse_reset()
{
	unsigned char param[2];

	if (psmouse_command(param, PSMOUSE_CMD_RESET_BAT))
		return -1;

	if (param[0] != PSMOUSE_RET_BAT && param[1] != PSMOUSE_RET_ID)
		return -1;

	return 0;
}


/*
 * psmouse_extensions() probes for any extensions to the basic PS/2 protocol
 * the mouse may have.
 */

static int psmouse_extensions()
{
	psmouse->vendor = "Generic";
	psmouse->name = "Mouse";
	psmouse->model = 0;

	return PSMOUSE_PS2;
}

/*
 * psmouse_probe() probes for a PS/2 mouse.
 */

static int psmouse_probe()
{
	unsigned char param[2];

/*
 * First, we check if it's a mouse. It should send 0x00 or 0x03
 * in case of an IntelliMouse in 4-byte mode or 0x04 for IM Explorer.
 */

	param[0] = 0xa5;

	if (psmouse_command(param, PSMOUSE_CMD_GETID))
		return -1;

	if (param[0] != 0x00 && param[0] != 0x03 && param[0] != 0x04)
		return -1;

/*
 * Then we reset and disable the mouse so that it doesn't generate events.
 */

	if (psmouse_command(NULL, PSMOUSE_CMD_RESET_DIS))
		warn("Failed to reset mouse\n");

/*
 * And here we try to determine if it has any extensions over the
 * basic PS/2 3-button mouse.
 */

	return psmouse->type = psmouse_extensions();
}

/*
 * Here we set the mouse resolution.
 */

static void psmouse_set_resolution()
{
	unsigned char param[1];

	if (!psmouse_resolution || psmouse_resolution >= 200)
		param[0] = 3;
	else if (psmouse_resolution >= 100)
		param[0] = 2;
	else if (psmouse_resolution >= 50)
		param[0] = 1;
	else if (psmouse_resolution)
		param[0] = 0;

        psmouse_command(param, PSMOUSE_CMD_SETRES);
}

/*
 * Here we set the mouse report rate.
 */

static void psmouse_set_rate()
{
	unsigned char rates[] = { 200, 100, 80, 60, 40, 20, 10, 0 };
	int i = 0;

	while (rates[i] > psmouse_rate) i++;
	psmouse_command(rates + i, PSMOUSE_CMD_SETRATE);
}

/*
 * psmouse_initialize() initializes the mouse to a sane state.
 */

static void psmouse_initialize()
{
	unsigned char param[2];

/*
 * We set the mouse report rate, resolution and scaling.
 */

	if (psmouse_max_proto != PSMOUSE_PS2) {
		psmouse_set_rate();
		psmouse_set_resolution();
		psmouse_command(NULL, PSMOUSE_CMD_SETSCALE11);
	}

/*
 * We set the mouse into streaming mode.
 */

	psmouse_command(param, PSMOUSE_CMD_SETSTREAM);
}

/*
 * psmouse_activate() enables the mouse so that we get motion reports from it.
 */

static void psmouse_activate()
{
	if (psmouse_command(NULL, PSMOUSE_CMD_ENABLE))
		warn("Failed to enable mouse\n");

	psmouse->state = PSMOUSE_ACTIVATED;
}

/*
 * psmouse_disconnect() closes and frees.
 */

void psmouse_disconnect()
{
	psmouse->state = PSMOUSE_CMD_MODE;

	if (psmouse->disconnect)
		psmouse->disconnect(psmouse);

	psmouse->state = PSMOUSE_IGNORE;
}

/*
 * psmouse_connect() is a callback from the serio module when
 * an unhandled serio port is found.
 */
int psmouse_connect()
{

	int probe_ok=0, i;
	memset(psmouse, 0, sizeof(struct psmouse));

	uinput_set_evbit(EV_KEY);
	uinput_set_evbit(EV_REL);

	uinput_set_keybit(BTN_LEFT);
	uinput_set_keybit(BTN_MIDDLE);
	uinput_set_keybit(BTN_RIGHT);

	uinput_set_relbit(REL_X);
	uinput_set_relbit(REL_Y);

	psmouse->state = PSMOUSE_CMD_MODE;

	for (i=0;i<100;i++) {
		if (DEBUG) printf("probing mouse\n");
		if (probe_ok==1) 
			break;
		probe_ok = psmouse_probe();
		if (DEBUG) printf("probe=%d\n",probe_ok);
	}

	if (probe_ok <= 0)
		return -1;

	sprintf(psmouse->devname, "%s %s %s",
		psmouse_protocols[psmouse->type], psmouse->vendor, psmouse->name);
	info("%s\n", psmouse->devname);

	psmouse_initialize();

	psmouse_activate();

	return 0;
}

static int psmouse_reconnect()
{
	int old_type;

	old_type = psmouse->type;

	psmouse->state = PSMOUSE_CMD_MODE;
	psmouse->type = psmouse->acking = psmouse->cmdcnt = psmouse->pktcnt = 0;
	if (psmouse->reconnect) {
	       if (psmouse->reconnect())
			return -1;
	} else if (psmouse_probe() != old_type)
		return -1;

	/* ok, the device type (and capabilities) match the old one,
	 * we can continue using it, complete intialization
	 */
	psmouse->type = old_type;
	psmouse_initialize();

	psmouse_activate();
	return 0;
}


psmouse.h
#ifndef _PSMOUSE_H
#define _PSMOUSE_H

#define PSMOUSE_CMD_SETSCALE11	0x00e6
#define PSMOUSE_CMD_SETRES	0x10e8
#define PSMOUSE_CMD_GETINFO	0x03e9
#define PSMOUSE_CMD_SETSTREAM	0x00ea
#define PSMOUSE_CMD_POLL	0x03eb
#define PSMOUSE_CMD_GETID	0x02f2
#define PSMOUSE_CMD_SETRATE	0x10f3
#define PSMOUSE_CMD_ENABLE	0x00f4
#define PSMOUSE_CMD_RESET_DIS	0x00f6
#define PSMOUSE_CMD_RESET_BAT	0x02ff

#define PSMOUSE_RET_BAT		0xaa
#define PSMOUSE_RET_ID		0x00
#define PSMOUSE_RET_ACK		0xfa
#define PSMOUSE_RET_NAK		0xfe

/* psmouse states */
#define PSMOUSE_CMD_MODE	0
#define PSMOUSE_ACTIVATED	1
#define PSMOUSE_IGNORE		2

struct psmouse;

struct psmouse {
	void *private;
	const char *vendor;
	const char *name;
	unsigned char cmdbuf[8];
	unsigned char packet[8];
	unsigned char cmdcnt;
	unsigned char pktcnt;
	unsigned char type;
	unsigned char model;
	unsigned char state;
	char acking;
	volatile char ack;
	char error;
	char devname[64];
	long long last;

	int (*reconnect)();
	void (*disconnect)();
};

#define PSMOUSE_PS2		1
#define PSMOUSE_PS2PP		2
#define PSMOUSE_PS2TPP		3
#define PSMOUSE_GENPS		4
#define PSMOUSE_IMPS		5
#define PSMOUSE_IMEX		6
#define PSMOUSE_SYNAPTICS 	7

extern int psmouse_command(unsigned char *param, int command);
extern int psmouse_reset(void);

extern void phys_reconnect();

extern int psmouse_smartscroll;
extern unsigned int psmouse_rate;
extern unsigned int psmouse_resetafter;

extern struct psmouse * const psmouse;


extern void uinput_set_evbit(int bit);
extern void uinput_set_ledbit(int bit);
extern void uinput_set_keybit(int bit);
extern void uinput_set_relbit(int bit);

extern void uinput_event(__u16 type, __u16 code, __s32 value);
#define uinput_report_key(code,value) uinput_event(EV_KEY, (code), (value))
#define uinput_report_rel(code,value) uinput_event(EV_REL, (code), (value))
#define uinput_report_abs(code,value) uinput_event(EV_ABS, (code), (value))
#define uinput_sync() uinput_event(EV_SYN, 0, 0)



#define warn(...) fprintf(stderr, "psmouse: (warning) " __VA_ARGS__)
#define err(...)  fprintf(stderr, "psmouse: ERROR " __VA_ARGS__)
#define info(...) printf("psmouse: " __VA_ARGS__)
#define notice(...) printf("psmouse: NOTE " __VA_ARGS__)
#define perr(msg)  do{fprintf(stderr, "psmouse: ERROR: "); perror(msg);}while(0)
#define perrx(msg) do{perr(msg); exit(1);}while(0)
#define pferrx() perrx(__FUNCTION__)
#define unlikely /* */

#endif /* _PSMOUSE_H */


opengalax.h
/*
 *   opengalax touchscreen daemon
 *   Copyright 2012 Pau Oliva Fora <pof@eslack.org>
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License
 *   as published by the Free Software Foundation; either version 2
 *   of the License, or (at your option) any later version.
 *
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <signal.h>
#include <linux/uinput.h>
#include <sys/stat.h>

#define XA_MAX 0xF
#define YA_MAX 0xF

#define XB_MAX	0x7F
#define YB_MAX	0x7F

#define X_AXIS_MAX (XA_MAX+1)*(XB_MAX+1)
#define Y_AXIS_MAX (YA_MAX+1)*(YB_MAX+1)

#define CMD_OK 0xFA
#define CMD_ERR 0xFE

#define PRESS 0x81
#define RELEASE 0x80

#define BTN1_RELEASE 0
#define BTN1_PRESS 1
#define BTN2_RELEASE 2
#define BTN2_PRESS 3

#define DEBUG 0

#define die(str, args...) do { \
	perror(str); \
	exit(EXIT_FAILURE); \
} while(0)

/* config file */
typedef struct {
	char serial_device[1024];
	char uinput_device[1024];
	int rightclick_enable;
	int rightclick_duration;
	int rightclick_range;
	int direction;
	int psmouse;
} conf_data;

typedef struct {
	int xmin;
	int xmax;
	int ymin;
	int ymax;
} calibration_data;

int fd_serial, fd_uinput;
struct uinput_user_dev uidev;
int use_psmouse;

/* configfile.c */
int create_config_file (char* file);
conf_data config_parse (void);
calibration_data calibration_parse (void);

/* functions.c */
int running_as_root (void);
int time_elapsed_ms (struct timeval *start, struct timeval *end, int ms); 
int configure_uinput (void);
int setup_uinput (void);
int setup_uinput_dev (const char *ui_dev);
int open_serial_port (const char *fd_device); 
int init_panel (); 
void initialize_panel (int sig);
void signal_handler (int sig);
void signal_installer (void);
int file_exists (char *file);
char* default_pid_file (void); 
int create_pid_file (void); 
int remove_pid_file (void);

/* psmouse.c */

void uinput_open(const char *uinput_dev_name); 
int psmouse_connect();
void uinput_create(); 
int phys_wait_for_input(int *ptimeout); 
void psmouse_interrupt(unsigned char data);
void uinput_destroy();
void uinput_close();
void psmouse_disconnect();


/*

init sequence:
 	send: f5 f3 0a f3 64 f3 c8 f4
 	read: fa fa fa fa fa fa fa fa

return values:
	0xFA == OK
	0xFE == ERROR

Example PDUs:
	81 0F 3D 05 06
	80 0F 3D 05 06

byte 0:
	0x80 == PRESS
	0x81 == RELEASE

byte 1:
	X axis value, from 0 to 0x0F
byte 2:
	X axis value, from 0 to 0x7F

byte 3:
	Y axis value, from 0 to 0x0F
byte 4:
	Y axis value, from 0 to 0x7F

*/


configfile.c
/*
 *   opengalax touchscreen daemon
 *   Copyright 2012 Pau Oliva Fora <pof@eslack.org>
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License
 *   as published by the Free Software Foundation; either version 2
 *   of the License, or (at your option) any later version.
 *
 */

#include "opengalax.h"

#define CONFIG_FILE "/etc/opengalax.conf"
#define MAXLEN 1024

static const conf_data default_config = {
	/* serial_device */ "/dev/serio_raw0",
	/* uinput_device */ "/dev/uinput",
	/* rightclick_enable */	0,
	/* rightclick_duration */ 350,
	/* rightclick_range */ 10,
	/* direction */ 0,
	/* psmouse */ 0,
};

static const calibration_data default_calibration = {
	/* xmin */ 0,
	/* xmax */ 2047,
	/* ymin */ 0,
	/* ymax */ 2047,
};

int create_config_file (char* file) {
	FILE* fd;

	fd = fopen(file, "w");
	if (fd == NULL)
		return 0;

	fprintf(fd, "# opengalax configuration file\n");
	fprintf(fd, "\n#### config data:\n");
	fprintf(fd, "serial_device=%s\n", default_config.serial_device);
	fprintf(fd, "uinput_device=%s\n", default_config.uinput_device);
	fprintf(fd, "rightclick_enable=%d\n", default_config.rightclick_enable);
	fprintf(fd, "rightclick_duration=%d\n", default_config.rightclick_duration);
	fprintf(fd, "rightclick_range=%d\n", default_config.rightclick_range);
	fprintf(fd, "# direction: 0 = normal, 1 = invert X, 2 = invert Y, 4 = swap X with Y\n");
	fprintf(fd, "direction=%d\n", default_config.direction);
	fprintf(fd, "# set psmouse=1 if you have a mouse connected into the same port\n");
	fprintf(fd, "# this usually requires i8042.nomux=1 and i8042.reset kernel parameters\n");
	fprintf(fd, "psmouse=%d\n", default_config.psmouse);
	fprintf(fd, "\n#### calibration data:\n");
	fprintf(fd, "# - values should range from 0 to 2047\n");
	fprintf(fd, "# - right/bottom must be bigger than left/top\n");
	fprintf(fd, "# left edge value:\n");
	fprintf(fd, "xmin=%d\n", default_calibration.xmin);
	fprintf(fd, "# right edge value:\n");
	fprintf(fd, "xmax=%d\n", default_calibration.xmax);
	fprintf(fd, "# top edge value:\n");
	fprintf(fd, "ymin=%d\n", default_calibration.ymin);
	fprintf(fd, "# bottom edge value:\n");
	fprintf(fd, "ymax=%d\n", default_calibration.ymax);
	fprintf(fd, "\n");

	fclose(fd);
	return 1;
}

conf_data config_parse (void) {

	char file[MAXLEN];
	char input[MAXLEN], temp[MAXLEN];
	FILE *fd;
	size_t len;
	conf_data config = default_config;

	sprintf( file, "%s", CONFIG_FILE);
	if (!file_exists(file)) {
		if (!create_config_file(file)) {
			fprintf (stderr,"Failed to create default config file: %s\n", file);
			exit (1);
		}
	}

	fd = fopen (file, "r");
	if (fd == NULL) {
		fprintf (stderr,"Could not open configuration file: %s\n", file);
		exit (1);
	}

	while ((fgets (input, sizeof (input), fd)) != NULL) {

		if ((strncmp ("serial_device=", input, 14)) == 0) {
			strncpy (temp, input + 14,MAXLEN-1);
			len=strlen(temp);
			temp[len-1]='\0';
			sprintf(config.serial_device, "%s", temp);
			memset (temp, 0, sizeof (temp));
		}

		if ((strncmp ("uinput_device=", input, 14)) == 0) {
			strncpy (temp, input + 14,MAXLEN-1);
			len=strlen(temp);
			temp[len-1]='\0';
			sprintf(config.uinput_device, "%s", temp);
			memset (temp, 0, sizeof (temp));
		}

		if ((strncmp ("rightclick_enable=", input, 18)) == 0) {
			strncpy (temp, input + 18,MAXLEN-1);
			len=strlen(temp);
			temp[len+1]='\0';
			config.rightclick_enable = atoi(temp);
		}

		if ((strncmp ("rightclick_duration=", input, 20)) == 0) {
			strncpy (temp, input + 20,MAXLEN-1);
			len=strlen(temp);
			temp[len+1]='\0';
			config.rightclick_duration = atoi(temp);
		}

		if ((strncmp ("rightclick_range=", input, 17)) == 0) {
			strncpy (temp, input + 17,MAXLEN-1);
			len=strlen(temp);
			temp[len+1]='\0';
			config.rightclick_range = atoi(temp);
		}

		if ((strncmp ("direction=", input, 10)) == 0) {
			strncpy (temp, input + 10,MAXLEN-1);
			len=strlen(temp);
			temp[len+1]='\0';
			config.direction = atoi(temp);
		}

		if ((strncmp ("psmouse=", input, 8)) == 0) {
			strncpy (temp, input + 8,MAXLEN-1);
			len=strlen(temp);
			temp[len+1]='\0';
			config.psmouse = atoi(temp);
		}
	}

	fclose(fd);
	return config;
}

calibration_data calibration_parse (void) {

	char file[MAXLEN];
	char input[MAXLEN], temp[MAXLEN];
	FILE *fd;
	size_t len;
	calibration_data calibration = default_calibration;

	sprintf( file, "%s", CONFIG_FILE);
	if (!file_exists(file)) {
		if (!create_config_file(file)) {
			fprintf (stderr,"Failed to create default config file: %s\n", file);
			exit (1);
		}
	}

	fd = fopen (file, "r");
	if (fd == NULL) {
		fprintf (stderr,"Could not open configuration file: %s\n", file);
		exit (1);
	}

	while ((fgets (input, sizeof (input), fd)) != NULL) {

		if ((strncmp ("xmin=", input, 5)) == 0) {
			strncpy (temp, input + 5,MAXLEN-1);
			len=strlen(temp);
			temp[len+1]='\0';
			calibration.xmin = atoi(temp);
		}

		if ((strncmp ("xmax=", input, 5)) == 0) {
			strncpy (temp, input + 5,MAXLEN-1);
			len=strlen(temp);
			temp[len+1]='\0';
			calibration.xmax = atoi(temp);
		}

		if ((strncmp ("ymin=", input, 5)) == 0) {
			strncpy (temp, input + 5,MAXLEN-1);
			len=strlen(temp);
			temp[len+1]='\0';
			calibration.ymin = atoi(temp);
		}

		if ((strncmp ("ymax=", input, 5)) == 0) {
			strncpy (temp, input + 5,MAXLEN-1);
			len=strlen(temp);
			temp[len+1]='\0';
			calibration.ymax = atoi(temp);
		}
	}

	fclose(fd);
	return calibration;
}
